docs(design/autoware-interfaces): add localization interface (#311)
autowarefoundation/autoware-documentation: docs/design/autoware-interfaces/components/localization.md

      - geometry_msg/msg/Quaternion orientation
    - double[36] covariance

### 3D-LiDAR Scanning

LiDAR scanning for NDT matching, published by the LiDAR sensor.

- sensor_msgs/msg/PointCloud2

The raw 3D-LiDAR data needs to be processed by the [point cloud pre-processing modules](../../autoware-architecture/sensing/data-types/point-cloud.md) before being used for localization.

### Automatic Initial pose

Start pose of ego, calculated from INS(Inertial navigation sensor) sensing data.

- geometry_msgs/msg/PoseWithCovarianceStamped
  - std_msgs/msg/Header header
  - geometry_msgs/msg/PoseWithCovariance pose
    - geometry_msgs/msg/Pose pose
      - geometry_msgs/msg/Point position
      - geometry_msg/msg/Quaternion orientation
    - double[36] covariance

When the initial pose is not set manually, the message can be used for automatic pose initialization.

Current Geographic coordinate of the ego, published by the GNSS sensor.

- sensor_msgs/msg/NavSatFix
  - std_msgs/msg/Header header
  - sensor_msgs/msg/NavSatStatus status
  - double latitude
  - double longitude
  - double altitude
  - double[9] position_covariance
  - unit8 position_covariance_type

> Current orientation of the ego, published by the IMU.

Reviewer: ```suggestion
Current orientation of the ego, published by the GNSS-INS.
```

> - autoware_sensing_msgs/msg/GnssinsOrientationStamped

Reviewer: ```suggestion
- autoware_sensing_msgs/msg/GnssInsOrientationStamped
```

  - std_msgs/Header header

>   - autoware_sensing_msgs/msg/GnssinsOrientation orientation

Reviewer: ```suggestion
  - autoware_sensing_msgs/msg/GnssInsOrientation orientation
```

    - geometry_msgs/Quaternion orientation
    - float32 rmse_rotation_x
    - float32 rmse_rotation_y
    - float32 rmse_rotation_z

### IMU Data

Current orientation, angular velocity and linear acceleration of ego, calculated from IMU sensing data.

- sensor_msgs/msg/Imu
  - std_msgs/msg/Header header
  - geometry_msgs/msg/Quaternion orientation
  - double[9] orientation_covariance
  - geometry_msgs/msg/Vector3 angular_velocity
  - double[9] angular_velocity_covariance
  - geometry_msgs/msg/Vector3 linear_acceleration
  - double[9] linear_acceleration_covariance

### Vehicle Velocity Status

Current velocity of the ego vehicle, published by the vehicle interface.

- autoware_auto_vehicle_msgs/msg/VelocityReport
  - std_msgs/msg/Header header;
  - float longitudinal_velocity;
  - float lateral_velocity;
  - float heading_rate;

Before the velocity input localization interface, module `vehicle_velocity_converter` converts message type `autoware_auto_vehicle_msgs/msg/VelocityReport` to `geometry_msgs/msg/TwistWithCovarianceStamped`.

## Outputs

> ### Vehicle pose

Reviewer: Position and velocity are now included in kinematic_state and are no longer output as localization output.

Author: if we also need pose, velocity and acceleration, we can use
PoseWithCovarianceStamped,TwistWithCovarianceStamped and AccelWithCovarianceStamped.

I don't know what will happen to the code in the future, but the localizationtion module of the current code does output these topics at the same time.

Reviewer: @cyn-liu sorry, I thought I had deleted the `/localization/pose_with_covariance` topic, but I see it is still there.

If that is the case, you can write Pose as output for documentation.
However, it would be helpful if you could specify that it is deprecated to use it as an output for localization, as it may be deleted in the future.

Author: @YamatoAndo  Thanks, your suggestion is very helpful to me.  I will specify in the document that the pose and twist topics have been included in the kinematics_ State, and will be abandoned in the future.

Author: @YamatoAndo  hello, I have a question:
The data type of the currently localization output `kinematic_state` topic is `nav_msgs/msg/Odometry`, only includes `pose` and `twist`, however,  the custom [KinematicState](https://github.com/autowarefoundation/autoware_msgs/blob/main/autoware_localization_msgs/msg/KinematicState.msg) includes 'pose', `twist` and `acceleration`, but is not used.

Reviewer: @cyn-liu 
You are correct, it is currently nav_msgs/msg/Odometry.
but will be changed to the custom [KinematicState](https://github.com/autowarefoundation/autoware_msgs/blob/main/autoware_localization_msgs/msg/KinematicState.msg) in the future.

I have created  a issue https://github.com/autowarefoundation/autoware.universe/issues/3024.

Author: OK, thanks @YamatoAndo 
I will update to custom [KinematicState](https://github.com/autowarefoundation/autoware_msgs/blob/main/autoware_localization_msgs/msg/KinematicState.msg) in this documentation.

# Localization

![Node diagram](images/Localization-Bus-ODD-Architecture.drawio.svg)

## Inputs

### PointCloud Map

Environment map created with point cloud, Published by the Map Server.

- sensor_msgs::msg::PointCloud2

### Initial Pose

Start pose of ego, Published by the User Interface.

> - geometry_msgs::msg::PoseCovarianceStamped

Reviewer: ```suggestion
- geometry_msgs::msg::PoseWithCovarianceStamped
```

### 3D-LiDAR Scanning

LiDAR scanning for NDT matching. Published by the LiDAR sensor.

- sensor_msgs::msg::PointCloud2

### GNSS

Current Geographic coordinate of the ego. Published by GNSS sensor.

> - sensor_msgs::msg::NavSatFix

Reviewer: ```suggestion
- sensor_msgs::msg::NavSatFix
- autoware_sensing_msgs::msg::GnssInsOrientationStamped
```
Incidentally, if we also need velocity and acceleration, we can use
TwistWithCovarianceStamped and AccelWithCovarianceStamped.
https://github.com/autowarefoundation/autoware_msgs/tree/main/autoware_sensing_msgs

# Localization

> ![Node diagram](images/Localization-Bus-ODD-Architecture.drawio.svg)

Reviewer 2: @xmfcx Did you recommend using Mermaid for component diagrams before? I don't remember that.

Reviewer 3: @kenji-miyake I've created this diagram in draw.io (diagrams.net) https://drive.google.com/file/d/1amCZ8sxCuY1isEVhGw99oKeIoqfV25Po/view?usp=sharing

So this was taken from there. But I am open to making these in Mermaid too.

Reviewer 2: I see, thank you.
I think draw.io is also acceptable for now.

## Inputs

### PointCloud Map

Environment map created with point cloud, Published by the Map Server.

- sensor_msgs::msg::PointCloud2

### Initial Pose

Start pose of ego, Published by the User Interface or GNSS.

> - geometry_msgs::msg::PoseStamped

Reviewer 4: It has been changed to a service `autoware_adapi_v1_msgs::srv::InitializeLocalization` to get result. Could you change to this? If necessary, you can define own messages for localization.
https://github.com/autowarefoundation/autoware_adapi_msgs

Author: thanks for your reply, I will change it soon.

Author: @isamu-takagi  hello:
After checking the latest code, I found my current tutorial is also acceptable, `rviz` published `initialpose` message is the `PoseWithCovarianceStamped` data type, autoware `ad_api_adaptors` package change it as a service to match the input of  the `pose_initializer` package.

Reviewer 4: Yes, that's right. More precisely, It's an input to the API, not localization, but I'll leave it up to you to use `PoseWithCovarianceStamped` for simplicity.
https://autowarefoundation.github.io/autoware.universe/main/system/default_ad_api/document/localization/

### 3D-LiDAR Scanning

LiDAR scanning for NDT matching. Published by the LiDAR sensor.

- sensor_msgs::msg::PointCloud2

### GNSS

Current Geographic coordinate of the ego. Published by GNSS sensor.

- sensor_msgs::msg::NavSatFix

### IMU

Current orientation, angular velocity and linear acceleration of ego. Published by IMU sensor.

- sensor_msgs::msg::Imu

### Vehicle Velocity Status

Current velocity of the ego vehicle. Published by the Vehicle Interface.

- autoware_auto_vehicle_msgs/msg/VelocityReport
  - std_msgs::msg::Header header;
  - float longitudinal_velocity;
  - float lateral_velocity;
  - float heading_rate;

## Outputs

### Vehicle pose

Current position and orientation of ego.

- geometry_msgs::msg::PoseWithCovarianceStamped

### Vehicle twist

Current line velocity and angular velocity of ego.

- geometry_msgs::msg::TwistWithCovarianceStamped

### Vehicle acceleration

Current acceleration of ego.

- geometry_msgs::msg::AccelWithCovarianceStamped

### Vehicle kinematic state

Current pose and twist of ego.

> - geometry_msgs::msg::Odometry

Reviewer 4: Odometry is in nav_msgs. Is it okay to be redundant with pose and twist?

Author: Yes, you're right.
The localization module  publishes `pose` and `twist` messages, but these two messages are include in `Kinematic State`, and other interfaces only subscribe to `Kinematic State` messages, so I will delete the pose and twist in this tutorial.
